app/models/transformation_mapping_item.rb

```diff
@@ -24,4 +31,36 @@ def destination_cluster
       errors.add(:destination, "EMS type of destination cluster or cloud tenant must be in: #{destination_types}")
     end
   end
+
+  def source_datastore
+    source_storage = source
+
+    cluster_storages = source.hosts # Get hosts using this source storage
+                             .collect(&:ems_cluster) # How many clusters does each host has
+                             .collect(&:storages).flatten # How many storages each host is mapped to that belong to the cluster
+
+    unless cluster_storages.include?(source_storage)
+      storage_types = VALID_SOURCE_DATASTORE_TYPES.join(', ')
+      errors.add(:source, "Source type must be in: #{storage_types}")
```

Comment:
What does the storage being in the set of cluster storages have to do with the storage type? Really confused

Reply:
When TransforamtionMappingItem of `:source_type => `storage'` is created, it needs a source storage and a destination storage.  The source storage must belong to the source cluster and the destination storage must belong to the destination cluster.

The way the storage model is setup, a storage in a TransformationMappingItem is associated to a host, the host can have one or more ems_clusters, and each cluster can have more that one storages. I hope this helps.

Reply:
Okay but you're not checking that the source is of type storage, so not sure how the error message "Source type must be in: Storage" helps ?

Reply:
I'm lost here too.  I don't know what this check is doing.    If source is a storage that's not a host's clusters storage, why are we saying it's must be a Storage?

Reply:
filtering mechanism changed.  now filtering on a TransformationMapping .